homeassistant/components/mqtt/mixins.py

```diff
@@ -0,0 +1,471 @@
+
+import voluptuous as vol
+
+from homeassistant.const import CONF_DEVICE, CONF_NAME
+from homeassistant.core import callback
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.dispatcher import (
+    async_dispatcher_connect,
+    async_dispatcher_send,
+)
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import ConfigType
+
+from . import CONF_TOPIC, DATA_MQTT, debug_info, publish
+from .const import (
+    ATTR_DISCOVERY_HASH,
+    ATTR_DISCOVERY_PAYLOAD,
+    ATTR_DISCOVERY_TOPIC,
+    CONF_QOS,
+    DEFAULT_PAYLOAD_AVAILABLE,
+    DEFAULT_PAYLOAD_NOT_AVAILABLE,
+    DOMAIN,
+    MQTT_CONNECTED,
+    MQTT_DISCONNECTED,
+)
+from .debug_info import log_messages
+from .discovery import (
+    MQTT_DISCOVERY_DONE,
+    MQTT_DISCOVERY_UPDATED,
+    clear_discovery_hash,
+    set_discovery_hash,
+)
+from .models import Message
+from .subscription import async_subscribe_topics, async_unsubscribe_topics
+from .util import valid_subscribe_topic
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_AVAILABILITY = "availability"
+CONF_AVAILABILITY_TOPIC = "availability_topic"
+CONF_PAYLOAD_AVAILABLE = "payload_available"
+CONF_PAYLOAD_NOT_AVAILABLE = "payload_not_available"
+CONF_JSON_ATTRS_TOPIC = "json_attributes_topic"
+CONF_JSON_ATTRS_TEMPLATE = "json_attributes_template"
+
+CONF_IDENTIFIERS = "identifiers"
+CONF_CONNECTIONS = "connections"
+CONF_MANUFACTURER = "manufacturer"
+CONF_MODEL = "model"
+CONF_SW_VERSION = "sw_version"
+CONF_VIA_DEVICE = "via_device"
+CONF_DEPRECATED_VIA_HUB = "via_hub"
+
+MQTT_AVAILABILITY_SINGLE_SCHEMA = vol.Schema(
+    {
+        vol.Exclusive(CONF_AVAILABILITY_TOPIC, "availability"): valid_subscribe_topic,
+        vol.Optional(
+            CONF_PAYLOAD_AVAILABLE, default=DEFAULT_PAYLOAD_AVAILABLE
+        ): cv.string,
+        vol.Optional(
+            CONF_PAYLOAD_NOT_AVAILABLE, default=DEFAULT_PAYLOAD_NOT_AVAILABLE
+        ): cv.string,
+    }
+)
+
+MQTT_AVAILABILITY_LIST_SCHEMA = vol.Schema(
+    {
+        vol.Exclusive(CONF_AVAILABILITY, "availability"): vol.All(
+            cv.ensure_list,
+            [
+                {
+                    vol.Optional(CONF_TOPIC): valid_subscribe_topic,
+                    vol.Optional(
+                        CONF_PAYLOAD_AVAILABLE, default=DEFAULT_PAYLOAD_AVAILABLE
+                    ): cv.string,
+                    vol.Optional(
+                        CONF_PAYLOAD_NOT_AVAILABLE,
+                        default=DEFAULT_PAYLOAD_NOT_AVAILABLE,
+                    ): cv.string,
+                }
+            ],
+        ),
+    }
+)
+
+MQTT_AVAILABILITY_SCHEMA = MQTT_AVAILABILITY_SINGLE_SCHEMA.extend(
+    MQTT_AVAILABILITY_LIST_SCHEMA.schema
+)
+
+
+def validate_device_has_at_least_one_identifier(value: ConfigType) -> ConfigType:
+    """Validate that a device info entry has at least one identifying value."""
+    if not value.get(CONF_IDENTIFIERS) and not value.get(CONF_CONNECTIONS):
```

Comment:
Personal opinion:

It's easier to read:

```python
if value.get(CONF_IDENTIFIERS) or value.get(CONF_CONNECTIONS):
    return value
raise vol.Invalid(
    "Device must have at least one identifying value in "
    "'identifiers' and/or 'connections'")
```

Reply:
Thanks, fixed.